UCT/IB: Support select gid by ndev

jeynmann opened the pull request:
## What
Support to select gid by network device name.

## Why ?

For `vxlan` and `bond` using the same IP address, ucx doesn't choose the gid index of `bond1` as we expect.
The gid index for `vxlan` and `bond` might be different for different physical nodes in 1 cluster.

``` sh
DEV     PORT    INDEX   GID                                     IPv4            VER     DEV
---     ----    -----   ---                                     ------------    ---     ---
mlx5_bond_0     1       0       fe80:0000:0000:0000:0ac0:ebff:fe79:391a                 v1      bond1
mlx5_bond_0     1       1       fe80:0000:0000:0000:0ac0:ebff:fe79:391a                 v2      bond1
mlx5_bond_0     1       2       0000:0000:0000:0000:0000:ffff:1e99:0aaf 30.153.10.175   v1      b3.vxlan.30100
mlx5_bond_0     1       3       0000:0000:0000:0000:0000:ffff:1e99:0aaf 30.153.10.175   v2      b3.vxlan.30100
mlx5_bond_0     1       4       0000:0000:0000:0000:0000:ffff:1e99:0aaf 30.153.10.175   v1      bond1
mlx5_bond_0     1       5       0000:0000:0000:0000:0000:ffff:1e99:0aaf 30.153.10.175   v2      bond1
```

## How ?
Add environment variable named GID_NDEV to allow UCT select gid by network device name.

JeynmannZ committed:
UCT/IB: support choose gid by ndev

JeynmannZ committed:
UCT/IB: support choose gid by ndev

assign gid ndev name after select gid

JeynmannZ committed:
UCT/IB: support choose gid by ndev

add test case for select gid by ndev name

JeynmannZ committed:
UCT/IB: support choose gid by ndev

 fix coverity + make format

leibin2014 commented:
/azp run

JeynmannZ committed:
UCT/IB: support choose gid by ndev

add debug info + explicit comment

JeynmannZ committed:
UCT/IB: support choose gid by ndev

  adjust format

rakhmets commented on src/uct/ib/base/ib_md.c:
```diff
@@ -1168,6 +1172,8 @@ ucs_status_t uct_ib_md_open_common(uct_ib_md_t *md,
                           UCT_MD_FLAG_ADVISE;
     md->reg_cost        = md_config->reg_cost;
     md->relaxed_order   = 0;
+    ucs_strncpy_zero(md->config.gid_ndev, md_config->gid_ndev,
+                     sizeof(md->config.gid_ndev));
```
I think it is better to store NULL in case of empty parameter. And then just pass `md->config.gid_ndev` to `uct_ib_iface_init_roce_gid_info`. It will simplify the code a little bit, because you can just call `uct_ib_device_select_gid_by_ndev(dev, port_num, cfg_gid_ndev, &iface->gid_info);`.

JeynmannZ committed:
UCT/IB: support choose gid by ndev

1. merge uct_ib_device_select_gid/uct_ib_device_select_gid_by_ndev
2. add uct_ib_device_get_ndev_name

JeynmannZ committed:
UCT/IB: support choose gid by ndev

1. add pointer to gid ndev

rakhmets commented on src/uct/ib/base/ib_device.h:
```diff
@@ -252,6 +252,7 @@ typedef struct {
     union ibv_gid              gid;
     uint8_t                    gid_index;    /* IB/RoCE GID index to use */
     uct_ib_roce_version_info_t roce_info;    /* For a RoCE port */
+    char                       ndev_name[IFNAMSIZ];
```
```suggestion
```

rakhmets commented on src/uct/ib/base/ib_device.c:
```diff
@@ -896,12 +899,16 @@ ucs_status_t uct_ib_device_select_gid(uct_ib_device_t *dev, uint8_t port_num,
             }
 
-            if ((roce_prio[prio_idx].ver         == gid_info_tmp.roce_info.ver) &&
+            if (((ndev_name == NULL) ||
+                    (strncmp(ndev_name, gid_info_tmp.ndev_name, IFNAMSIZ) == 0)) &&
+                (roce_prio[prio_idx].ver         == gid_info_tmp.roce_info.ver) &&
```
```suggestion
            if ((roce_prio[prio_idx].ver         == gid_info_tmp.roce_info.ver) &&
```

Please add at the beginning of the loop body:
```C
if (ucs_string_is_empty(ndev_name)) {
    uct_ib_device_get_roce_ndev_name(dev, port_num, i, ndev_name_tmp, sizeof(ndev_name_tmp))); 
    if (strcmp(ndev_name, ndev_name_tmp) != 0) {
        continue;
    }
}
```
,  you will also need local variable `char ndev_name_tmp[IFNAMSIZ]`.

rakhmets commented on src/uct/ib/base/ib_device.h:
```diff
@@ -385,6 +388,10 @@ uct_ib_device_create_ah_cached(uct_ib_device_t *dev,
                                struct ibv_ah_attr *ah_attr, struct ibv_pd *pd,
                                const char *usage, struct ibv_ah **ah_p);
 
+ucs_status_t uct_ib_device_get_ndev_name(const char *dev_name, uint8_t port_num,
```
Why do you need extra function here. I guess you can use `uct_ib_device_get_roce_ndev_name`.

rakhmets commented on src/uct/ib/base/ib_md.h:
```diff
@@ -88,6 +88,8 @@ typedef struct uct_ib_md_ext_config {
     } odp;
 
     unsigned long            gid_index;    /**< IB GID index to use */
+    char                     *gid_ndev;    /**< IB GID network device to use */
+    char                     ndev_name[IFNAMSIZ]; /**< IB GID network device to use */
```
```suggestion
    char                     gid_ndev[IFNAMSIZ]; /**< IB GID network device to use */
```

rakhmets commented on src/uct/ib/base/ib_md.c:
```diff
@@ -1169,6 +1173,14 @@ ucs_status_t uct_ib_md_open_common(uct_ib_md_t *md,
     md->reg_cost        = md_config->reg_cost;
     md->relaxed_order   = 0;
 
+    if (strnlen(md_config->gid_ndev, 1) == 0) {
+        md->config.gid_ndev = NULL;
+    } else {
+        ucs_strncpy_zero(md->config.ndev_name, md_config->gid_ndev,
+                         sizeof(md->config.ndev_name));
+        md->config.gid_ndev = md->config.ndev_name;
+    }
```
```suggestion
     ucs_strncpy_zero(md->config.gid_ndev, md_config->gid_ndev,
                     sizeof(md->config.gid_ndev));
```

JeynmannZ committed:
UCT/IB: support choose gid by ndev

simplifize code according to the comments

JeynmannZ committed:
UCT/IB: support choose gid by ndev

fix gtest

JeynmannZ committed:
UCT/IB: support choose gid by ndev

simplifize code according to the comments

rakhmets commented on src/uct/ib/base/ib_device.c:
```diff
@@ -889,6 +891,15 @@ ucs_status_t uct_ib_device_select_gid(uct_ib_device_t *dev, uint8_t port_num,
         for (i = 0; i < gid_tbl_len; i++) {
+            if (!ucs_string_is_empty(ndev_name)) {
+                status = uct_ib_device_get_roce_ndev_name(
+                    dev, port_num, i, ndev_name_tmp, sizeof(ndev_name_tmp));
```
```suggestion
                        dev, port_num, i, ndev_name_tmp, sizeof(ndev_name_tmp));
```

rakhmets commented on src/uct/ib/base/ib_md.c:
```diff
@@ -1169,6 +1173,9 @@ ucs_status_t uct_ib_md_open_common(uct_ib_md_t *md,
     md->relaxed_order   = 0;
 
+    ucs_strncpy_zero(md->config.gid_ndev, md_config->gid_ndev,
+	                 sizeof(md->config.gid_ndev));
```
```suggestion
	             sizeof(md->config.gid_ndev));
```

JeynmannZ committed:
UCT/IB: support choose gid by ndev

adjust formats according to the comments

JeynmannZ committed:
UCT/IB: support choose gid by ndev

reset unused changes

svc-nixl commented:
Can one of the admins verify this patch?

gleon99 commented:
Hi @jeynmann - we're cleaning up the UCX PR backlog and plan to close this PR in 7 days. Please let us know if you'd like to keep it open.
